src/ToData.purs

```diff
@@ -1,37 +1,158 @@
 module ToData
-  ( class ToData
+  ( AnotherDay(..)
+  , Day(..)
+  , Tree(..)
+  , class ConstrIndex
+  , class CountedConstrIndex
+  , class ToData
+  , class ToDataArgs
+  , class ToDataWithIndex
+  , constrIndex
+  , countedConstrIndex
+  , defaultConstrIndex
+  , genericToData
   , toData
+  , toDataArgs
+  , toDataWithIndex
   ) where
 
 import Prelude
 
 import Data.Array as Array
+import Data.Array ((..))
 import Data.BigInt (BigInt)
 import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
 import Data.Foldable (class Foldable)
-import Data.List (List)
+import Data.Generic.Rep as G
 import Data.Map (Map)
+import Data.List (List)
 import Data.Map as Map
 import Data.Maybe (Maybe(Just, Nothing))
 import Data.Profunctor.Strong ((***))
 import Data.Ratio (Ratio, denominator, numerator)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol)
+import Data.Tuple (Tuple)
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+-- Classes
+
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  constrIndex :: Proxy a -> Map String Int
+
+class ToDataWithIndex a ci where
+  toDataWithIndex :: ConstrIndex ci => Proxy ci -> a -> PlutusData
+
+-- As explained in https://harry.garrood.me/blog/write-your-own-generics/ this
+-- is just a neat pattern that flattens a skewed Product of Products
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+-- Default constructor indices
+class CountedConstrIndex (a :: Type) where
+  countedConstrIndex :: Proxy a -> Int -> Map String Int -> Map String Int
```

Comment:
I think it would be super helpful if you could add a docstring for `ConstrIndex`, `ToDataWithIndex` and `CountedConstrIndex` perhaps with a very basic example for the readers.